src/workerd/api/sql.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#pragma once
+
+#include <workerd/jsg/jsg.h>
+#include "basics.h"
+
+namespace workerd::api {
+
+typedef kj::OneOf<kj::Array<const byte>, kj::String, double> ValueBind;
+class SqlDatabase;
+
+class SqlResult final: public jsg::Object {
+public:
+  SqlResult(SqliteDatabase::Query& query);
+
+  kj::ArrayPtr<jsg::Dict<SqliteDatabase::Query::ValueOwned>> getResults(jsg::Lock& js);
+
+  JSG_RESOURCE_TYPE(SqlResult, CompatibilityFlags::Reader flags) {
+    JSG_READONLY_PROTOTYPE_PROPERTY(results, getResults);
+  }
+
+private:
+  kj::Vector<jsg::Dict<SqliteDatabase::Query::ValueOwned>> results;
+};
+
+class SqlPreparedStatement final: public jsg::Object {
+public:
+  SqlPreparedStatement(jsg::Ref<SqlDatabase>&& sqlDb, SqliteDatabase::Statement&& statement);
+
+  jsg::Ref<SqlResult> run(kj::Maybe<kj::Array<ValueBind>> bindValuesOptional);
+
+  JSG_RESOURCE_TYPE(SqlPreparedStatement, CompatibilityFlags::Reader flags) {
+    JSG_METHOD(run);
+  }
+
+private:
+  void visitForGc(jsg::GcVisitor& visitor) {
+    visitor.visit(sqlDatabase);
+  }
+
+  jsg::Ref<SqlDatabase> sqlDatabase;
+  SqliteDatabase::Statement statement;
+};
+
+class SqlDatabase final: public jsg::Object {
+public:
+  friend class SqlPreparedStatement;
+  SqlDatabase(SqliteDatabase& sqlite);
+
+  jsg::Ref<SqlResult> exec(jsg::Lock& js, kj::String query, bool admin, kj::Maybe<kj::Array<ValueBind>> bindValuesOptional);
+  jsg::Ref<SqlPreparedStatement> prepare(jsg::Lock& js, kj::String query, bool admin);
+
+  JSG_RESOURCE_TYPE(SqlDatabase, CompatibilityFlags::Reader flags) {
+    JSG_METHOD(exec);
+    JSG_METHOD(prepare);
+  }
+
+private:
+  static int authorize(
+    void* userdata,
+    int actionCode,
+    const char* param1,
+    const char* param2,
+    const char* dbName,
+    const char* triggerName);
+
+  void onError(const char* errStr);
+
+  SqliteDatabase& sqlite;
```

Comment:
Is the `SqliteDatabase` instance here guaranteed to always outlive the JS `SqlDatabase` instance here? If I do funny things like stick the `SqlDatabase` into the global scope and access it there, or pass it into a `waitUntil`, is there any risk that the `SqlliteDatabase` instance will be freed first leading to a UAF here?

Reply:
Ah good catch. What I would actually like to do is just bind it as a member of DurableObjectStorage (SqlDatabase's lifetime should be tied to the lifetime of DurableObjectStorage). Is there a way to delete jsg objects (I'm assuming not), or bind members like this? It seems odd to have SqlDatabase implement visitForGc to keep DurableObjectStorage alive.

Reply:
If you don't want the `SqlDatabase` instance to keep the `DurableObjectStorage` instance alive you can use a weakref pattern like we do in a few other places. Specifically, something like the following...

```
class WeakRef : public kj::Refcounted {
public:
  WeakRef(SqliteDatabase* ref) : ref(ref) {}
  kj::Own addRef() { return kj::addRef(*this); }
  kj::Maybe<SqliteDatabase&> tryGet() {
    return ref.map([](auto& ptr) { return *ptr; });
  }
  void reset() { ref = nullptr; }
private:
  kj::Maybe<SqliteDatabase*> ref;
};
```

Then, whatever owns your `SqliteDatabase` instance creates a initial local instance of the WeakRef...

```
auto ref = kj::refcounted<WeakRef>(&sqliteDatabase)
```

And passes a ref (`ref->addRef()`) into your JSG object.

When your owning object is destroyed, it calls `ref->reset()` to clear the held pointer and the JSG object starts seeing a nullptr.

That has limitations if you're mixing kj heap and js heap objects, tho, so tread carefully. @kentonv likely has thoughts on that case